File: examples/skyrmion_excitation.py
import matplotlib.pyplot as plt
import numpy as np

from mumax5 import Ferromagnet, Grid, World
from mumax5.util import neelskyrmion, show_field


# NUMERICAL PARAMETERS RELEVANT FOR THE SPECTRUM ANALYSIS
fmax = 50E9           # maximum frequency (in Hz) of the sinc pulse
T = 2E-9              # simulation time (longer -> better frequency resolution)
dt = 1 / (2 * fmax)  # the sample time (Nyquist theorem taken into account)
t0 = 1 / fmax
d = 100E-9            # circle diameter
nx = 32               # number of cells


# create the world
grid_size = (nx, nx, 1)
cell_size = (d / nx, d / nx, 1E-9)

world = World(cell_size)

# create the ferromagnet
geometry_func = lambda x, y, z: (x - 4.84E-8) ** 2 +  (y -  4.84E-8) ** 2 <= (0.5 * d) ** 2

# mask = mask.reshape((1, 1, 32, 32)) # the bug is not cause by our mask, so its good
# magnet = Ferromagnet(world, Grid(size=grid_size), geometry=geometry_func)
magnet = Ferromagnet(world, Grid(size=grid_size))
magnet.geometry
magnet.msat = 1E6
magnet.aex = 15E-12

mask = np.zeros(shape=(1, 1, 32, 32))
magnet.ku1 = lambda t: 1E6 * (1 + 0.01 * np.sinc(2 * fmax * (t - t0))), mask
# magnet.ku1 = 0

magnet.idmi = 3E-3
magnet.anisU = (0, 0, 1)
magnet.alpha = 0.001

# set and relax the initial magnetization
magnet.magnetization = neelskyrmion(position=magnet.center,
                                    radius=0.5 * d,
                                    charge=-1,
                                    polarization=1)
magnet.minimize()

timepoints = np.linspace(0, T, int(T / dt))
outputquantities = {
    "mx": lambda: magnet.magnetization.average()[0],
    "my": lambda: magnet.magnetization.average()[1],
    "mz": lambda: magnet.magnetization.average()[2],
    "e_total": magnet.total_energy,
    "e_exchange": magnet.exchange_energy,
    "e_zeeman": magnet.zeeman_energy,
    "e_demag": magnet.demag_energy,
}

# --- RUN THE SOLVER ---

world.timesolver.timestep

output = world.timesolver.solve(timepoints, outputquantities)

# --- PLOT THE OUTPUT DATA ---

plt.figure(figsize=(10, 8))
for key in ["mx", "my", "mz"]:
    plt.plot(output["time"], output[key], label=key)
plt.legend()
plt.title("Mumax5 - Python")
plt.show()

# FAST FOURIER TRANSFORM
dm     = np.array(output["mz"]) - output["mz"][0]   # average magnetization deviaton
spectr = np.abs(np.fft.fft(dm))         # the absolute value of the FFT of dm
freq   = np.linspace(0, 1/dt, len(dm))  # the frequencies for this FFT

# PLOT THE SPECTRUM
plt.plot(freq/1e9, spectr)
plt.xlim(0,fmax/1e9)
plt.ylabel("Spectrum (a.u.)")
plt.xlabel("Frequency (GHz)")
plt.title("Mumax5 - Python")
plt.show()

show_field(magnet.magnetization)


plt.subplot(211)
for key in ["mx", "my", "mz"]:
    plt.plot(output["time"], output[key], label=key)
plt.legend()

plt.subplot(212)
for key in ["e_total", "e_exchange", "e_zeeman", "e_demag"]:
    plt.plot(timepoints, output[key], label=key)
plt.legend()

plt.show()

File: src/cmd/skyrmion_excitation.cpp
#include "examples.hpp"

#include <fstream>
#include <iostream>
#include <limits>
#include <cmath>

#include "dynamicequation.hpp"
#include "ferromagnet.hpp"
#include "field.hpp"
#include "grid.hpp"
#include "mumaxworld.hpp"
#include "timesolver.hpp"
#include "torque.hpp"

void skyrmion_excitation() {
    std::cout << "Skyrmion Excitation (Vairable Anisotropy)" << std::endl;

    real fmax = 50E9;  // maximum frequency(in Hz) of the sinc pulse
    real T = 2E-9;  // simulation time(longer->better frequency resolution) 
    real dt =
        1 / (2 * fmax);  // sample time(Nyquist theorem taken into account)
    real t0 = 1 / fmax;


    real length = 100E-9, width = 100E-9, thickness = 1E-9;
    int3 n = make_int3(32, 32, 1);
    real3 cellsize{length / n.x, width / n.y, thickness / n.z};
    std::string ferromagnet_name = "";

    MumaxWorld mWorld(cellsize);
    Grid mGrid(n);
    auto magnet = mWorld.addFerromagnet(mGrid, ferromagnet_name);

    magnet->msat.set(1E6);
    magnet->aex.set(15E-12);
    magnet->alpha.set(0.001);
    magnet->anisU.set(real3{0, 0, 1});
    magnet->idmi.set(3E-3);

    auto sinc = [](real x) -> real {
      if (std::abs(x) <= std::numeric_limits<real>::epsilon())
        return 1.0;
      else
        return sin(x) / (x);
    };

    auto Ku1t = [&sinc, fmax, T, t0](real t) -> real {
      return 1E6 * (1.0 + 0.01 * sinc(2 * M_PI * fmax * (t - t0)));
    };

    magnet->ku1.addTimeDependentTerm(Ku1t); // do we need a mask??

    magnet->minimize();

    // how to create a Neel skyrmion here??
    // m = neelskyrmion(-1, 1)

    real start = 0, stop = T;
    int n_timepoints = static_cast<int>(1 + (stop - start) / dt);

    std::string out_file_path = "./spin_wave_m.csv";
    std::ofstream magn_csv(out_file_path);
    magn_csv << "t,mx,my,mz,bx" << std::endl;

    for (int i = 0; i < n_timepoints; i++) {
      mWorld.timesolver().run(dt);
      auto m = magnet->magnetization()->average();
      magn_csv << mWorld.time() << "," << m[0] << "," << m[1] << "," << m[2]
               << "," << Ku1t(i * dt) << "," << std::endl;
    }
}
